torch/_decomp/decompositions.py

```diff
@@ -1547,6 +1547,48 @@ def maybe_mask(vals, length, range_max, adaptive, dim):
     return ret / (length_h * length_w)
 
 
+@register_decomposition(aten.index_copy)
+def index_copy(x, dim, index, tensor):
+    return x.clone().index_copy_(dim, index, tensor)
+
+
+@register_decomposition(aten.index_add)
+def index_add(x, dim, index, tensor, *, alpha=1):
+    return x.clone().index_add_(dim, index, tensor, alpha=alpha)
+
+
+@register_decomposition(aten.index_select, disable_meta=True)
+def index_select(x, dim, index):
+    dim = utils.canonicalize_dims(x.ndim, dim)
+    utils.check(index.ndim <= 1, lambda: f"Index should have dimension 1 or 0 (got {index.ndim})")
+    utils.check(utils.is_integer_dtype(index.dtype), lambda: "Expected dtype int32 or int64 for index")
+    # Treat scalars as elements of \R^1
+    if x.ndim == 0:
+        return x.unsqueeze(0)[index].squeeze(0)
+    idx = (slice(None),) * dim + (index,)
+    return x[idx]
+
+
+@register_decomposition(aten.index_copy_)
+def index_copy_(x, dim, index, tensor):
+    dim = utils.canonicalize_dims(x.ndim, dim)
+    utils.check(index.ndim <= 1, lambda: f"Index should have dimension 1 or 0 (got {index.ndim})")
+    utils.check(utils.is_integer_dtype(index.dtype), lambda: "Expected dtype int32 or int64 for index")
+    idx = (slice(None),) * dim + (index,)
+    x[idx] = tensor
+    return x
+
+
+@register_decomposition(aten.index_add_)
+def index_add_(x, dim, index, tensor, *, alpha=1):
+    dim = utils.canonicalize_dims(x.ndim, dim)
+    utils.check(index.ndim <= 1, lambda: f"Index should have dimension 1 or 0 (got {index.ndim})")
+    utils.check(utils.is_integer_dtype(index.dtype), lambda: "Expected dtype int32 or int64 for index")
+    idx = (slice(None),) * dim + (index,)
+    x[idx] += alpha * tensor
```

Comment:
Isn't `x[idx] +=` the same as `index_add_`?

Reply:
The general `x[idx] +=` is implemented via `index_put_(..., accumulate=True)`. Same for `x[idx] =` but with `accumulate=False`.

Reply:
Nb. The comment above is not true. See https://github.com/pytorch/pytorch/pull/85002#issuecomment-1247188844 and https://github.com/pytorch/pytorch/pull/85002#issuecomment-1248524492